File: cli/src/freepod/vars.py
"""`freepod var`: a deployment's runtime configuration.

Values are strings on the wire. A sensitive var is write-only — the platform
returns it with no `value` at all — so an entry carrying no value means "leave
this one alone", which is what makes `var list --json` output safe to submit
back through `var set -f -`.
"""

from __future__ import annotations

import json
import sys
from pathlib import Path
from typing import Any, Dict, List, Optional, Tuple

import click

from . import FreepodError, UsageError
from .api import ApiClient
from .table import GAP

#: The only phase that currently exists. Leaves room for build vars.
PHASE = "runtime"

#: Stands in for a value the platform will not return.
HIDDEN = "<hidden>"

COLUMNS = ("KEY", "VALUE")


def _base(user_id: int, deployment_id: str) -> str:
    return f"/api/users/{user_id}/deployments/{deployment_id}/vars/{PHASE}"


def read(api: ApiClient, user_id: int, deployment_id: str) -> Dict[str, Any]:
    """The deployment's vars envelope: `{"vars": {...}, "pending": bool}`."""
    body = api.get_json(_base(user_id, deployment_id))
    if not isinstance(body, dict):
        raise FreepodError(f"unexpected vars response: {body!r}")
    return body


def write(
    api: ApiClient, user_id: int, deployment_id: str, entries: Dict[str, Any]
) -> Dict[str, Any]:
    """Merge `entries` into the deployment's vars."""
    response = api.request(
        "PATCH", _base(user_id, deployment_id), json={"vars": entries}
    )
    if not response.is_success:
        raise _write_error(response)
    return response.json()


def remove(api: ApiClient, user_id: int, deployment_id: str, keys: List[str]) -> None:
    """Delete keys. Removing one that is not set is a no-op, not an error."""
    for key in keys:
        response = api.delete(f"{_base(user_id, deployment_id)}/{key}")
        if not response.is_success:
            raise _write_error(response)


def _write_error(response) -> FreepodError:
    detail = ""
    try:
        body = response.json()
        detail = body.get("detail") if isinstance(body, dict) else ""
    except ValueError:
        detail = response.text.strip()[:300]
    return FreepodError(detail or f"the platform refused the write: HTTP {response.status_code}")


# --------------------------------------------------------------------------
# Input
# --------------------------------------------------------------------------


def parse_assignments(
    arguments: List[str], *, interactive: bool, prompt=click.prompt
) -> Dict[str, str]:
    """`KEY=VALUE` pairs; a bare `KEY` is prompted for without echo.

    Prompting is what keeps a secret out of the shell history, so a bare key
    with nowhere to prompt is a usage error rather than an empty value.
    """
    values: Dict[str, str] = {}
    for argument in arguments:
        key, separator, value = argument.partition("=")
        key = key.strip()
        if not key:
            raise UsageError(f"'{argument}' is not a KEY=VALUE pair")
        if separator:
            values[key] = value
            continue
        if not interactive:
            raise UsageError(
                f"no value given for {key}, and there is no terminal to prompt on.\n"
                f"  Pass {key}=VALUE, or -f FILE, or pipe the wire shape into -f -."
            )
        values[key] = prompt(f"Value for {key}", hide_input=True)
    return values


def load_entries(source: str) -> Dict[str, Any]:
    """Read vars from a file, or from stdin when `source` is `-`.

    Accepts the platform's own wire shape (so `var list --json` round-trips)
    or plain `KEY=VALUE` lines.
    """
    if source == "-":
        text = sys.stdin.read()
        origin = "standard input"
    else:
        path = Path(source)
        if not path.is_file():
            raise UsageError(f"{source}: no such file")
        text = path.read_text(encoding="utf-8")
        origin = source

    stripped = text.lstrip()
    if stripped.startswith("{"):
        try:
            document = json.loads(text)
        except json.JSONDecodeError as exc:
            raise UsageError(f"{origin}: not valid JSON ({exc})") from exc
        return _entries_from_wire(document, origin)
    return {key: {"value": value} for key, value in _entries_from_lines(text, origin).items()}


def _entries_from_wire(document: Any, origin: str) -> Dict[str, Any]:
    if not isinstance(document, dict):
        raise UsageError(f"{origin}: expected an object")
    # Both the envelope and a bare map, because one is what `--json` prints and
    # the other is what a person writing the file by hand would produce.
    vars_ = document.get("vars", document)
    if not isinstance(vars_, dict):
        raise UsageError(f"{origin}: 'vars' is not an object")

    entries: Dict[str, Any] = {}
    for key, entry in vars_.items():
        if isinstance(entry, str) or entry is None:
            entries[key] = {"value": entry}
            continue
        if not isinstance(entry, dict):
            raise UsageError(f"{origin}: {key} is neither a string nor an object")
        # Only the two fields a write carries. `updated_at`/`updated_by` come
        # back on a read and are the platform's to set.
        kept: Dict[str, Any] = {}
        if "value" in entry:
            kept["value"] = entry["value"]
        if "sensitive" in entry:
            kept["sensitive"] = entry["sensitive"]
        entries[key] = kept
    return entries


def _entries_from_lines(text: str, origin: str) -> Dict[str, str]:
    values: Dict[str, str] = {}
    for number, line in enumerate(text.splitlines(), start=1):
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            continue
        key, separator, value = stripped.partition("=")
        if not separator or not key.strip():
            raise UsageError(f"{origin}:{number}: expected KEY=VALUE, got {line!r}")
        values[key.strip()] = value
    return values


def schema_declares(deployment: Dict[str, Any], key: str) -> bool:
    """Whether the deployment's template declares `key`, and so owns its
    sensitivity."""
    template = deployment.get("desired_template") or {}
    schema = template.get("values_schema_json") or {}
    properties = schema.get("properties") if isinstance(schema, dict) else None
    return isinstance(properties, dict) and key in properties


def mark_sensitive(
    entries: Dict[str, Any], deployment: Dict[str, Any]
) -> Tuple[Dict[str, Any], List[str]]:
    """Flag every entry sensitive, except where the schema already decides.

    Returns the entries and the keys left alone, which the caller warns about.
    """
    marked: Dict[str, Any] = {}
    declared: List[str] = []
    for key, entry in entries.items():
        if schema_declares(deployment, key):
            declared.append(key)
            marked[key] = entry
            continue
        marked[key] = {**entry, "sensitive": True}
    return marked, declared


# --------------------------------------------------------------------------
# Output
# --------------------------------------------------------------------------


def rows(payload: Dict[str, Any]) -> List[Tuple[str, str]]:
    entries = payload.get("vars") or {}
    return [
        (key, HIDDEN if "value" not in entry else str(entry.get("value", "")))
        for key, entry in sorted(entries.items())
    ]


def render_table(payload: Dict[str, Any]) -> str:
    body = rows(payload)
    if not body:
        return ""
    widths = [max(len(str(cell)) for cell in column) for column in zip(COLUMNS, *body)]
    lines = [GAP.join(head.ljust(width) for head, width in zip(COLUMNS, widths)).rstrip()]
    lines += [
        GAP.join(str(cell).ljust(width) for cell, width in zip(row, widths)).rstrip()
        for row in body
    ]
    return "\n".join(lines)


def pending_count(
    api: ApiClient, user_id: int, deployment: Dict[str, Any]
) -> Optional[int]:
    """How many vars a rollout would change, or None if it cannot be told.

    Needs the applied release's snapshot, which the deployment read does not
    inline; fetched only when the deployment reports something pending.
    """
    if not deployment.get("pending"):
        return 0
    head = deployment.get("vars") or {}
    applied = deployment.get("applied_release") or {}
    number = applied.get("number")
    if number is None:
        return len(head)
    path = (
        f"/api/users/{user_id}/deployments/{deployment['id']}/releases/{number}"
    )
    response = api.get(path)
    if not response.is_success:
        return None
    running = (response.json() or {}).get("vars") or {}
    changed = {
        key for key in set(head) | set(running) if _entry(head, key) != _entry(running, key)
    }
    return len(changed)


def _entry(entries: Dict[str, Any], key: str) -> Any:
    entry = entries.get(key)
    if not isinstance(entry, dict):
        return None
    # A sensitive var carries no value either side, so it compares by its
    # timestamp: that is what changes when it is rewritten.
    return entry.get("value"), entry.get("sensitive"), entry.get("updated_at")

File: cli-rust/src/vars.rs
//! `freepod var`: a deployment's runtime configuration.
//!
//! Mirrors `vars.py`. Values are strings on the wire. A sensitive var is
//! write-only — the platform returns it with no `value` at all — so an entry
//! carrying no value means "leave this one alone", which is what makes
//! `var list --json` output safe to submit back through `var set -f -`.

use std::collections::HashMap;
use std::io::Read;

use serde_json::{json, Map, Value};

use crate::api::{ApiResponse, ApiClient};
use crate::errors::{freepod, usage, Error, Result};
use crate::table::{format_time, render, value_str, BLANK};

/// The only phase that currently exists. Leaves room for build vars.
pub const PHASE: &str = "runtime";

/// Stands in for a value the platform will not return.
pub const HIDDEN: &str = "<hidden>";

const COLUMNS: [&str; 4] = ["KEY", "VALUE", "UPDATED", "BY"];

fn base(user_id: u64, deployment_id: &str) -> String {
    format!("/api/users/{user_id}/deployments/{deployment_id}/vars/{PHASE}")
}

/// The deployment's vars envelope: `{"vars": {...}, "pending": bool}`.
pub async fn read(api: &mut ApiClient, user_id: u64, deployment_id: &str) -> Result<Value> {
    let body = api.get_json(&base(user_id, deployment_id), None).await?;
    if !body.is_object() {
        return Err(freepod(format!("unexpected vars response: {body}")));
    }
    Ok(body)
}

/// Merge `entries` into the deployment's vars.
pub async fn write(
    api: &mut ApiClient,
    user_id: u64,
    deployment_id: &str,
    entries: &Value,
) -> Result<Value> {
    let response = api
        .request(
            "PATCH",
            &base(user_id, deployment_id),
            None,
            Some(&json!({ "vars": entries })),
            None,
        )
        .await?;
    if !response.is_success() {
        return Err(write_error(&response));
    }
    response.decode()
}

/// Delete keys. Removing one that is not set is a no-op, not an error.
pub async fn remove(
    api: &mut ApiClient,
    user_id: u64,
    deployment_id: &str,
    keys: &[String],
) -> Result<()> {
    for key in keys {
        let response = api
            .delete(&format!("{}/{}", base(user_id, deployment_id), key))
            .await?;
        if !response.is_success() {
            return Err(write_error(&response));
        }
    }
    Ok(())
}

fn write_error(response: &ApiResponse) -> Error {
    let text = response.text();
    let detail = match serde_json::from_str::<Value>(&text) {
        Ok(Value::Object(obj)) => obj
            .get("detail")
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string(),
        Ok(_) => String::new(),
        Err(_) => text.trim().chars().take(300).collect(),
    };
    if detail.is_empty() {
        freepod(format!("the platform refused the write: HTTP {}", response.status))
    } else {
        freepod(detail)
    }
}

// --------------------------------------------------------------------------
// Input
// --------------------------------------------------------------------------

/// `KEY=VALUE` pairs; a bare `KEY` is prompted for without echo.
///
/// Prompting is what keeps a secret out of the shell history, so a bare key
/// with nowhere to prompt is a usage error rather than an empty value.
pub fn parse_assignments(
    arguments: &[String],
    interactive: bool,
) -> Result<HashMap<String, String>> {
    let mut values: HashMap<String, String> = HashMap::new();
    for argument in arguments {
        let (key, value, has_separator) = match argument.find('=') {
            Some(idx) => (
                argument[..idx].trim().to_string(),
                argument[idx + 1..].to_string(),
                true,
            ),
            None => (argument.trim().to_string(), String::new(), false),
        };
        if key.is_empty() {
            return Err(usage(format!("'{argument}' is not a KEY=VALUE pair")));
        }
        if has_separator {
            values.insert(key, value);
            continue;
        }
        if !interactive {
            return Err(usage(format!(
                "no value given for {key}, and there is no terminal to prompt on.\n  \
                 Pass {key}=VALUE, or -f FILE, or pipe the wire shape into -f -."
            )));
        }
        let secret = crate::prompt::prompt_secret(&format!("Value for {key}"))?;
        values.insert(key, secret);
    }
    Ok(values)
}

/// Read vars from a file, or from stdin when `source` is `-`.
///
/// Accepts the platform's own wire shape (so `var list --json` round-trips)
/// or plain `KEY=VALUE` lines.
pub fn load_entries(source: &str) -> Result<Value> {
    let (text, origin) = if source == "-" {
        let mut text = String::new();
        std::io::stdin()
            .read_to_string(&mut text)
            .map_err(|e| freepod(format!("could not read standard input: {e}")))?;
        (text, "standard input".to_string())
    } else {
        let path = std::path::Path::new(source);
        if !path.is_file() {
            return Err(usage(format!("{source}: no such file")));
        }
        let text = std::fs::read_to_string(path)
            .map_err(|e| freepod(format!("cannot read {source}: {e}")))?;
        (text, source.to_string())
    };

    if text.trim_start().starts_with('{') {
        let document: Value = serde_json::from_str(&text)
            .map_err(|e| usage(format!("{origin}: not valid JSON ({e})")))?;
        entries_from_wire(&document, &origin)
    } else {
        let mut obj = Map::new();
        for (key, value) in entries_from_lines(&text, &origin)? {
            obj.insert(key, json!({ "value": value }));
        }
        Ok(Value::Object(obj))
    }
}

fn entries_from_wire(document: &Value, origin: &str) -> Result<Value> {
    let doc = match document.as_object() {
        Some(o) => o,
        None => return Err(usage(format!("{origin}: expected an object"))),
    };
    // Both the envelope and a bare map, because one is what `--json` prints
    // and the other is what a person writing the file by hand would produce.
    let vars_ = match doc.get("vars") {
        Some(v) => v,
        None => document,
    };
    let vars_ = match vars_.as_object() {
        Some(o) => o,
        None => return Err(usage(format!("{origin}: 'vars' is not an object"))),
    };

    let mut entries = Map::new();
    for (key, entry) in vars_ {
        match entry {
            Value::String(_) | Value::Null => {
                entries.insert(key.clone(), json!({ "value": entry }));
            }
            Value::Object(fields) => {
                // Only the two fields a write carries. `updated_at`/`updated_by`
                // come back on a read and are the platform's to set.
                let mut kept = Map::new();
                if let Some(v) = fields.get("value") {
                    kept.insert("value".to_string(), v.clone());
                }
                if let Some(s) = fields.get("sensitive") {
                    kept.insert("sensitive".to_string(), s.clone());
                }
                entries.insert(key.clone(), Value::Object(kept));
            }
            _ => {
                return Err(usage(format!(
                    "{origin}: {key} is neither a string nor an object"
                )))
            }
        }
    }
    Ok(Value::Object(entries))
}

fn entries_from_lines(text: &str, origin: &str) -> Result<Vec<(String, String)>> {
    let mut values: Vec<(String, String)> = Vec::new();
    for (index, line) in text.lines().enumerate() {
        let stripped = line.trim();
        if stripped.is_empty() || stripped.starts_with('#') {
            continue;
        }
        let (key, value) = match stripped.find('=') {
            Some(idx) => (stripped[..idx].trim().to_string(), stripped[idx + 1..].to_string()),
            None => {
                return Err(usage(format!(
                    "{origin}:{}: expected KEY=VALUE, got '{line}'",
                    index + 1
                )))
            }
        };
        if key.is_empty() {
            return Err(usage(format!(
                "{origin}:{}: expected KEY=VALUE, got '{line}'",
                index + 1
            )));
        }
        values.push((key, value));
    }
    Ok(values)
}

/// Whether the deployment's template declares `key`, and so owns its
/// sensitivity.
pub fn schema_declares(deployment: &Value, key: &str) -> bool {
    let template = deployment
        .get("desired_template")
        .filter(|v| v.is_object())
        .unwrap_or(&Value::Null);
    let schema = template
        .get("values_schema_json")
        .filter(|v| v.is_object())
        .unwrap_or(&Value::Null);
    match schema.get("properties") {
        Some(Value::Object(properties)) => properties.contains_key(key),
        _ => false,
    }
}

/// Flag every entry sensitive, except where the schema already decides.
///
/// Returns the keys left alone, which the caller warns about.
pub fn mark_sensitive(entries: &mut Value, deployment: &Value) -> Vec<String> {
    let mut declared = Vec::new();
    if let Some(obj) = entries.as_object_mut() {
        for (key, entry) in obj.iter_mut() {
            if schema_declares(deployment, key) {
                declared.push(key.clone());
                continue;
            }
            if let Some(e) = entry.as_object_mut() {
                e.insert("sensitive".to_string(), Value::Bool(true));
            }
        }
    }
    declared
}

// --------------------------------------------------------------------------
// Output
// --------------------------------------------------------------------------

/// Who last wrote it, by email. The id is in `--json` and not here: it
/// identifies an account to a program, not to a person.
pub fn author(entry: &Value) -> String {
    let Some(writer) = entry.get("updated_by").and_then(|v| v.as_object()) else {
        return BLANK.to_string();
    };
    writer
        .get("email")
        .and_then(|v| v.as_str())
        .filter(|s| !s.is_empty())
        .map(String::from)
        .unwrap_or_else(|| BLANK.to_string())
}

/// One row per var, sorted by key, headers first.
pub fn rows(payload: &Value) -> Vec<Vec<String>> {
    let mut table: Vec<Vec<String>> = vec![COLUMNS.iter().map(|s| s.to_string()).collect()];
    if let Some(vars) = payload.get("vars").and_then(|v| v.as_object()) {
        let mut keys: Vec<&String> = vars.keys().collect();
        keys.sort();
        for key in keys {
            let entry = &vars[key];
            let value = match entry.get("value") {
                None => HIDDEN.to_string(),
                Some(v) => value_str(v),
            };
            table.push(vec![
                key.clone(),
                value,
                format_time(entry.get("updated_at").unwrap_or(&Value::Null)),
                author(entry),
            ]);
        }
    }
    table
}

/// The rendered table, or empty where the deployment has no vars.
pub fn render_table(payload: &Value) -> String {
    let body = rows(payload);
    if body.len() <= 1 {
        return String::new();
    }
    render(&body)
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn parse_assignments_splits_on_the_first_separator() {
        let values = parse_assignments(&["A=1".into(), "B=x=y".into()], false).unwrap();
        assert_eq!(values.get("A").map(String::as_str), Some("1"));
        assert_eq!(values.get("B").map(String::as_str), Some("x=y"));
    }

    #[test]
    fn parse_assignments_rejects_an_empty_key() {
        assert!(matches!(
            parse_assignments(&["=v".into()], false),
            Err(Error::Usage(_))
        ));
    }

    #[test]
    fn parse_assignments_bare_key_needs_a_terminal() {
        assert!(matches!(
            parse_assignments(&["SECRET".into()], false),
            Err(Error::Usage(_))
        ));
    }

    #[test]
    fn load_entries_reads_the_wire_shape() {
        let value = load_entries_from_text(
            r#"{"vars": {"A": "1", "B": {"value": "2", "sensitive": true, "updated_at": "x"}}}"#,
        )
        .unwrap();
        let obj = value.as_object().unwrap();
        assert_eq!(obj.get("A"), Some(&json!({ "value": "1" })));
        assert_eq!(
            obj.get("B"),
            Some(&json!({ "value": "2", "sensitive": true }))
        );
    }

    #[test]
    fn load_entries_reads_key_value_lines() {
        let value = load_entries_from_text("# comment\nA=1\n\nB=two words\n").unwrap();
        let obj = value.as_object().unwrap();
        assert_eq!(obj.get("A"), Some(&json!({ "value": "1" })));
        assert_eq!(obj.get("B"), Some(&json!({ "value": "two words" })));
    }

    #[test]
    fn load_entries_rejects_a_bad_line() {
        assert!(matches!(
            load_entries_from_text("A=1\nnot a pair\n"),
            Err(Error::Usage(_))
        ));
    }

    /// `load_entries` reads files or stdin; the tests feed it text directly.
    fn load_entries_from_text(text: &str) -> Result<Value> {
        let stripped = text.trim_start();
        if stripped.starts_with('{') {
            let document: Value = serde_json::from_str(text)
                .map_err(|e| usage(format!("test: not valid JSON ({e})")))?;
            entries_from_wire(&document, "test")
        } else {
            let mut obj = Map::new();
            for (key, value) in entries_from_lines(text, "test")? {
                obj.insert(key, json!({ "value": value }));
            }
            Ok(Value::Object(obj))
        }
    }

    #[test]
    fn schema_declares_reads_the_desired_template() {
        let deployment = json!({
            "desired_template": {
                "values_schema_json": { "properties": { "hostname": {} } }
            }
        });
        assert!(schema_declares(&deployment, "hostname"));
        assert!(!schema_declares(&deployment, "other"));
        assert!(!schema_declares(&json!({}), "hostname"));
        assert!(!schema_declares(&json!({ "desired_template": null }), "hostname"));
    }

    #[test]
    fn mark_sensitive_skips_schema_owned_keys() {
        let deployment = json!({
            "desired_template": {
                "values_schema_json": { "properties": { "hostname": {} } }
            }
        });
        let mut entries = json!({ "hostname": { "value": "a" }, "SECRET": { "value": "b" } });
        let declared = mark_sensitive(&mut entries, &deployment);
        assert_eq!(declared, vec!["hostname"]);
        assert_eq!(
            entries.get("SECRET").and_then(|v| v.get("sensitive")),
            Some(&json!(true))
        );
        assert!(entries.get("hostname").and_then(|v| v.get("sensitive")).is_none());
    }

    #[test]
    fn rows_hide_sensitive_values() {
        let payload = json!({
            "vars": {
                "A": { "value": "1", "updated_by": { "email": "a@x" } },
                "B": { "sensitive": true }
            }
        });
        let table = rows(&payload);
        assert_eq!(table[1][1], "1");
        assert_eq!(table[2][1], HIDDEN);
        assert_eq!(table[1][3], "a@x");
    }
}
